src/main/java/io/strimzi/kafka/quotas/VolumeSource.java

```diff
@@ -58,64 +61,72 @@ public VolumeSource(Admin admin, VolumeObserver volumeObserver, int timeout, Tim
 
     @Override
     public void run() {
-        log.info("Updating cluster volume usage.");
-        CompletableFuture<Map<Integer, Map<String, LogDirDescription>>> logDirsPerBrokerPromise = new CompletableFuture<>();
-        log.debug("Attempting to describe cluster");
-        admin.describeCluster().nodes().whenComplete((nodes, throwable) -> {
-            if (throwable != null) {
-                log.error("error while describing cluster", throwable);
-                logDirsPerBrokerPromise.completeExceptionally(throwable);
-            } else {
-                if (log.isDebugEnabled()) {
-                    log.debug("Successfully described cluster: " + nodes);
-                }
-                //Deliberately stay on the adminClient thread as the next thing we do is another admin API call
-                onDescribeClusterSuccess(nodes, logDirsPerBrokerPromise);
-            }
-        });
-
         try {
+            log.info("Updating cluster volume usage.");
+            CompletableFuture<VolumeUsageResult> volumeUsageResultPromise = new CompletableFuture<>();
+            log.debug("Attempting to describe cluster");
+            admin.describeCluster().nodes().whenComplete((nodes, throwable) -> {
+                if (throwable != null) {
+                    log.error("Error while describing cluster", throwable);
+                    volumeUsageResultPromise.complete(failure(VolumeSourceObservationStatus.DESCRIBE_CLUSTER_ERROR, throwable));
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Successfully described cluster: " + nodes);
+                    }
+                    //Deliberately stay on the adminClient thread as the next thing we do is another admin API call
```

Comment:
I think in practice it will only be executed by the thread calling `whenComplete` if the future is already completed when invoked. We have tweaked the wording to clarify, but I think the intent is still valid to remain on the adminClient thread as the next thing we do is make another admin client call.